File: packages/snow_draw_core/lib/draw/render/tasks/frame_render_plan.dart
import 'package:meta/meta.dart';

import '../../models/camera_state.dart';
import '../../utils/list_equality.dart';
import 'render_tasks.dart';

/// Immutable frame-level rendering plan produced by core.
@immutable
class FrameRenderPlan {
  const FrameRenderPlan({
    required this.tasks,
    required this.camera,
    required this.scaleFactor,
    this.localeTag,
  });

  /// Ordered frame tasks to execute for the frame.
  final List<FrameRenderTask> tasks;

  /// Camera transform snapshot.
  final CameraState camera;

  /// Effective scale factor used for world/canvas transforms.
  final double scaleFactor;

  /// Optional locale hint for text layout/rendering.
  final String? localeTag;

  static const empty = FrameRenderPlan(
    tasks: <FrameRenderTask>[],
    camera: CameraState.initial,
    scaleFactor: 1,
  );

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is FrameRenderPlan &&
          other.camera == camera &&
          other.scaleFactor == scaleFactor &&
          other.localeTag == localeTag &&
          listEquals(other.tasks, tasks);

  @override
  int get hashCode =>
      Object.hash(listHash(tasks), camera, scaleFactor, localeTag);
}

File: packages/snow_draw_rust_engine/src/draw/render/tasks/frame_render_plan.rs
#![allow(dead_code)]

use crate::draw::models::camera_state::CameraState;

use super::render_tasks::FrameRenderTask;

/// Immutable frame-level rendering plan produced by the engine.
#[derive(Clone, Debug, PartialEq)]
pub struct FrameRenderPlan {
    /// Ordered frame tasks to execute for the frame.
    pub tasks: Vec<FrameRenderTask>,
    /// Camera transform snapshot.
    pub camera: CameraState,
    /// Effective scale factor used for world/canvas transforms.
    pub scale_factor: f64,
    /// Optional locale hint for text layout/rendering.
    pub locale_tag: Option<String>,
}

impl FrameRenderPlan {
    pub fn new(
        tasks: Vec<FrameRenderTask>,
        camera: CameraState,
        scale_factor: f64,
        locale_tag: Option<String>,
    ) -> Self {
        Self {
            tasks,
            camera,
            scale_factor,
            locale_tag,
        }
    }

    pub fn empty() -> Self {
        Self::new(Vec::new(), CameraState::INITIAL, 1.0, None)
    }
}

impl Default for FrameRenderPlan {
    fn default() -> Self {
        Self::empty()
    }
}
